src/index.js

```diff
@@ -86,116 +108,146 @@ export default function nodeResolve ( options = {} ) {
 			// disregard entry module
 			if ( !importer ) return null;
 
-			if (options.browser && browserMapCache[importer]) {
-				const resolvedImportee = resolve( dirname( importer ), importee );
-				const browser = browserMapCache[importer];
-				if (browser[importee] === false || browser[resolvedImportee] === false) {
-					return ES6_BROWSER_EMPTY;
-				}
-				if (browser[importee] || browser[resolvedImportee] || browser[resolvedImportee + '.js'] || browser[resolvedImportee + '.json']) {
-					importee = browser[importee] || browser[resolvedImportee] || browser[resolvedImportee + '.js'] || browser[resolvedImportee + '.json'];
-				}
-			}
-
-
-			const parts = importee.split( /[/\\]/ );
-			let id = parts.shift();
-
-			if ( id[0] === '@' && parts.length ) {
-				// scoped packages
-				id += `/${parts.shift()}`;
-			} else if ( id[0] === '.' ) {
-				// an import relative to the parent dir of the importer
-				id = resolve( importer, '..', importee );
-			}
-
-			if (only && !only.some(pattern => pattern.test(id))) return null;
-
-			let disregardResult = false;
-			let packageBrowserField = false;
-			const extensions = options.extensions || DEFAULT_EXTS;
-
-			const resolveOptions = {
-				basedir: dirname( importer ),
-				packageFilter ( pkg, pkgPath ) {
-					const pkgRoot = dirname( pkgPath );
-					if (options.browser && typeof pkg[ 'browser' ] === 'object') {
-						packageBrowserField = Object.keys(pkg[ 'browser' ]).reduce((browser, key) => {
-							const resolved = pkg[ 'browser' ][ key ] === false ? false : resolve( pkgRoot, pkg[ 'browser' ][ key ] );
-							browser[ key ] = resolved;
-							if ( key[0] === '.' ) {
-								const absoluteKey = resolve( pkgRoot, key );
-								browser[ absoluteKey ] = resolved;
-								if ( !extname(key) ) {
-									extensions.reduce( ( browser, ext ) => {
-										browser[ absoluteKey + ext ] = browser[ key ];
-										return browser;
-									}, browser );
-								}
-							}
-							return browser;
-						}, {});
+			return Promise.resolve(this.cache ? this.cache.get(`${importee}|${importer}`) : false)
+				.then((cache) => {
+					// If the importer, importee or any package.json files have changed then invalidate the cache:
+					return cache && cache.stats.reduce((promise, cacheEntry) => {
+						return promise.then(stillValid => stillValid && compareFileCacheEntry(cacheEntry), Promise.resolve(true));
```

Comment:
So the gap here, is that a file that isn't tracked by the cache can in theory invalidated the resolver cache. Although perhaps this is enough of an edge case not to be an issue.

Example scenarios would be adding `file.mjs` when `file.js` was loaded previously.

Reply:
My understanding is that just adding `file.mjs` would not change how the module is resolved - unless other files around it changed. Is this correct?

The `cache.stats` list will store the importer, importee stats - plus any possible `package.json` files that could influence them, plus any `index.js` files it could potentially hit. In other words adding files which might intercept the loader algorithm would invalidate the cache precisely because we check every file that could intercept the loader algorithm.

Reply:
Node.js will resolve `file | file.mjs | file.js | file.json | file.node`. So if you're loading one of the later ones, then adding an earlier entry will take precedence regardless of the stat of the previous file.

Similarly it will resolve `dir/package.json>main | dir/index.mjs | dir/index.js | dir/index.json | dir/index.node` so that a similar precedence applies there.

Then there is also the case of `node_modules/pkgX` checking `node_modules/pkgX/node_modules/pkgY` before `node_modules/pkgY` so that if we have cached the lookup there to `node_modules/pkgY` then adding a `node_modules/pkgX/node_modules/pkgY` won't be seen.

For these reasons the cache approach I'd recommend is always re-running the resolver, but with a stat cache and package.json cache with a timed invalidation (at the risk of repeating myself, but admittedly I know I've skipped the above logic until now).

Reply:
I ran a quick test with the `node-resolve` function to see what files it looked for. The importer was `path/to/foo.js` which imports `fetch` (which is found at `node_modules/fetch/fetch.js`). `node-resolve` will look for the following files:
 - `path/to/node_modules/fetch`
 - `path/to/node_modules/fetch.js`
 - `path/to/node_modules/fetch/package.json`
 - `path/to/node_modules/fetch/index`
 - `path/to/node_modules/fetch/index.js`
 - `path/to/fetch`
 - `path/to/fetch.js`
 - `path/to/fetch/package.json`
 - `path/to/fetch/index`
 - `path/to/fetch/index.js`
 - `path/node_modules/fetch`
 - `path/node_modules/fetch.js`
 - `path/node_modules/fetch/package.json`
 - `path/node_modules/fetch/index`
 - `path/node_modules/fetch/index.js`
 - `path/fetch`
 - `path/fetch.js`
 - `path/fetch/package.json`
 - `path/fetch/index`
 - `path/fetch/index.js`
 - `node_modules/fetch`
 - `node_modules/fetch.js`
 - `node_modules/fetch/package.json` (found this file!)
 - `node_modules/fetch/fetch.js` (found this file!)

`node-resolves` checks every file with `isFile()` and `readFile()` which we instrument. Every one of those files gets statted as part of this code. Only 2 out of the 24 files exist. The ones that don't are still stored in the cache as `mtimeMs: 0, size: 0`. So if I later go and add `node_modules/fetch.js` to intercept that path, then the stat cache for that file will not match and the whole cache is invalidated.

In other words, any file `node-resolve` attempts to read - we know about, and we store some info about, to later check.

Reply:
That definitely catches everything then, but my worry in this scenario is we are now storing all the stats per individual resolution, and then restatting them for each resolution right?

So we now have a lot of cached data to load and work with, as well as still doing all the stats. So where is the performance benefit then? Have you run the numbers on this?